sources/metricdog/src/service_check.rs

```diff
@@ -0,0 +1,159 @@
+use crate::error::{self, Result};
+use log::trace;
+use snafu::ResultExt;
+use std::process::Command;
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) struct ServiceHealth {
+    /// Whether or not the service is healthy.
+    pub(crate) is_healthy: bool,
+    /// In the event of an unhealthy service, the service's exit code (if found).
+    pub(crate) exit_code: Option<i32>,
+}
+
+pub(crate) trait ServiceCheck {
+    /// Checks the given service to see if it is healthy.
+    fn check(&self, service_name: &str) -> Result<ServiceHealth>;
+}
+
+pub(crate) struct SystemdCheck {}
+
+impl ServiceCheck for SystemdCheck {
+    fn check(&self, service_name: &str) -> Result<ServiceHealth> {
+        if is_ok(service_name)? {
+            return Ok(ServiceHealth {
+                is_healthy: true,
+                exit_code: None,
+            });
+        }
+        Ok(ServiceHealth {
+            is_healthy: false,
+            exit_code: parse_service_exit_code(service_name)?,
+        })
+    }
+}
+
+struct Outcome {
+    exit: i32,
+    stdout: String,
+}
+
+impl Outcome {
+    fn is_exit_true(&self) -> bool {
+        self.exit == 0
+    }
+}
+
+fn systemctl(args: &[&str]) -> Result<Outcome> {
+    trace!("calling systemctl with '{:?}'", args);
+    let output = Command::new("systemctl")
+        .args(args)
+        .output()
+        .context(error::Command {
```

Comment:
```suggestion
        .with_context(|| error::Command {
```
To avoid the iterator and copies in non-error cases.

Reply:
Fixed in [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)